client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java

```diff
@@ -522,11 +527,156 @@ protected void set(K key, V value) {
+         return removedSomething;
+      }
+   }
+
+   private CloseableIterator<Entry<K, V>> castIterator(CloseableIterator iterator) {
+      return iterator;
+   }
+
+   private boolean removeEntry(Map.Entry<K, V> entry) {
+      K key = entry.getKey();
+      VersionedValue<V> versionedValue = RemoteCacheImpl.this.getVersioned(key);
+      if (versionedValue != null && entry.getValue().equals(versionedValue.getValue())) {
+         return RemoteCacheImpl.this.removeWithVersion(key, versionedValue.getVersion());
+      }
+      return false;
+   }
+
+   private class EntrySet extends AbstractCollection<Map.Entry<K, V>> implements CloseableIteratorSet<Entry<K, V>> {
+
+      @Override
+      public CloseableIterator<Entry<K, V>> iterator() {
+         return new RemovableCloseableIterator<>(castIterator(retrieveEntries(null, batchSize)),
+               RemoteCacheImpl.this::removeEntry);
+      }
+
+      @Override
+      public CloseableSpliterator<Entry<K, V>> spliterator() {
+         return Closeables.spliterator(iterator(), Long.MAX_VALUE, Spliterator.NONNULL | Spliterator.CONCURRENT |
+               Spliterator.DISTINCT);
+      }
+
+      @Override
+      public int size() {
+         return RemoteCacheImpl.this.size();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+         Map.Entry entry = toEntry(o);
+         if (entry != null) {
+            V value = RemoteCacheImpl.this.get(entry.getKey());
+            return value != null && value.equals(entry.getValue());
+         }
+         return false;
+      }
+
+      @Override
+      public boolean remove(Object o) {
+         Map.Entry<K, V> entry = toEntry(o);
+         if (entry != null) {
+            return RemoteCacheImpl.this.removeEntry(entry);
+         }
+         return false;
+      }
+
+      private Map.Entry<K, V> toEntry(Object obj) {
+         if (obj instanceof Map.Entry) {
+            return (Map.Entry) obj;
+         } else {
+            return null;
+         }
+      }
+   }
+
+   private class ValuesCollection extends AbstractCollection<V> implements CloseableIteratorCollection<V> {
+
+      @Override
+      public CloseableIterator<V> iterator() {
+         return new CloseableIteratorMapper<>(new RemovableCloseableIterator<>(castIterator(retrieveEntries(null, batchSize)),
+               RemoteCacheImpl.this::removeEntry), Map.Entry::getValue);
+      }
+
+      @Override
+      public CloseableSpliterator<V> spliterator() {
+         return Closeables.spliterator(iterator(), Long.MAX_VALUE, Spliterator.NONNULL | Spliterator.CONCURRENT);
+      }
+
+      @Override
+      public int size() {
+         return RemoteCacheImpl.this.size();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+         return RemoteCacheImpl.this.containsValue(o);
+      }
+
+      @Override
+      public boolean remove(Object o) {
+         Objects.requireNonNull(o);
+         try (CloseableIterator<V> iter = iterator()) {
```

Comment:
Uuh, my bad. I've mistaken that for the iterator.remove() operation. Sorry about that.